internal/cmd/update.go

```diff
@@ -0,0 +1,395 @@
+				fs:             afero.NewOsFs(),
+				osF:            func() string { return runtime.GOOS },
+				versionF:       func() string { return version.Version },
+			}
+			return updater.Run(ctx, force, coderURL)
+		},
+	}
+
+	cmd.Flags().BoolVar(&force, "force", false, "do not prompt for confirmation")
+	cmd.Flags().StringVar(&coderURL, "coder", "", "coder instance against which to match version")
+
+	return cmd
+}
+
+type getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+func (u *updater) Run(ctx context.Context, force bool, coderURLString string) error {
+	// Check under following directories and warn if coder binary is under them:
+	//   * C:\Windows\
+	//   * homebrew prefix
+	//   * coder assets root (/var/tmp/coder)
+	var pathBlockList = []string{
+		`C:\Windows\`,
+		`/var/tmp/coder`,
+	}
+	brewPrefixCmd, err := u.execF(ctx, "brew", "--prefix")
+	if err == nil { // ignore errors if homebrew not installed
+		pathBlockList = append(pathBlockList, strings.TrimSpace(string(brewPrefixCmd)))
+	}
+
+	for _, prefix := range pathBlockList {
+		if HasFilePathPrefix(u.executablePath, prefix) {
+			return clog.Fatal(
+				"cowardly refusing to update coder binary",
+				clog.BlankLine,
+				clog.Causef("executable path %q is under blocklisted prefix %q", u.executablePath, prefix))
+		}
+	}
+
+	currentBinaryStat, err := u.fs.Stat(u.executablePath)
+	if err != nil {
+		return clog.Fatal("preflight: cannot stat current binary", clog.Causef(err.Error()))
+	}
+
+	if currentBinaryStat.Mode().Perm()&0222 == 0 {
+		return clog.Fatal("preflight: missing write permission on current binary")
+	}
+
+	var coderURL *url.URL
+	if coderURLString == "" {
+		coderURL, err = getCoderConfigURL()
+		if err != nil {
+			return clog.Fatal(
+				"Unable to automatically determine coder URL",
+				clog.Causef(err.Error()),
+				clog.BlankLine,
+				clog.Tipf("use --coder <url> to specify coder URL"),
+			)
+		}
+	} else {
+		coderURL, err = url.Parse(coderURLString)
+		if err != nil {
+			return clog.Fatal("invalid coder URL", err.Error())
+		}
+	}
+
+	desiredVersion, err := getAPIVersionUnauthed(u.httpClient, *coderURL)
+	if err != nil {
+		return clog.Fatal("fetch api version", clog.Causef(err.Error()))
+	}
+
+	clog.LogInfo(fmt.Sprintf("Coder instance at %q reports version %s", coderURL.String(), desiredVersion.String()))
+	clog.LogInfo(fmt.Sprintf("Current version of coder-cli is %s", version.Version))
+
+	if currentVersion, err := semver.StrictNewVersion(u.versionF()); err == nil {
+		if desiredVersion.Compare(currentVersion) == 0 {
+			clog.LogInfo("Up to date!")
+			return nil
+		}
+	}
+
+	if !force {
+		label := fmt.Sprintf("Do you want to download version %s instead", desiredVersion)
+		if _, err := u.confirmF(label); err != nil {
+			return clog.Fatal("failed to confirm update", clog.Tipf(`use "--force" to update without confirmation`))
+		}
+	}
+
+	downloadURL := makeDownloadURL(desiredVersion, u.osF(), runtime.GOARCH)
```

Comment:
Adding a `--tag` flag to explicitly specify a version to download could be helpful. I could imagine us publishing an RC and wanting to `coder update --tag` to that. Another option would be `coder update --next` to use prereleases.

Reply:
Done, but `coder update --version <version_spec>` instead.